### PR TITLE
Upgradle

### PR DESCRIPTION
**Pull request by paplorinc**
Updating the build to latest Gradle, Spock and JUnit in preparation for a Config Cache fix.
Please review [commit-by-commit](https://github.com/gradle/test-retry-gradle-plugin/pull/103/commits) for context.

### CODE REVIEW

File: plugin/build.gradle.kts [hunk 1]
```diff
@@ -33,15 +33,16 @@ val plugin: Configuration by configurations.creating
...
-    plugin("org.ow2.asm:asm:9.0")
+    plugin("org.ow2.asm:asm:9.1")
 
     testImplementation(gradleTestKit())
     testImplementation(localGroovy())
-    testImplementation("org.spockframework:spock-core:1.3-groovy-2.5")
+    testImplementation("org.spockframework:spock-core:2.0-groovy-3.0")
+    testImplementation("org.spockframework:spock-junit4:2.0-groovy-3.0")
     testImplementation("net.sourceforge.nekohtml:nekohtml:1.9.22")
-    testImplementation("org.ow2.asm:asm:8.0.1")
```
Comment (paplorinc) [inline plugin/build.gradle.kts:42]: why was the test version different from the plugin version (line 36)?

File: plugin/build.gradle.kts [hunk 2]
```diff
@@ -57,12 +58,14 @@ shadowJar.configure {
...
-tasks.getByName("jar").dependsOn(shadowJar)
+tasks.getByName("jar").apply {
+    enabled = false
+    dependsOn(shadowJar)
+}
 
 gradlePlugin {
     plugins {
-        create("testRetry") {
+        register("testRetry") {
```
Comment (paplorinc) [inline plugin/build.gradle.kts:68]: is it okay to make these lazy?

File: plugin/build.gradle.kts [hunk 3]
```diff
@@ -57,12 +58,14 @@ shadowJar.configure {
...
             id = "org.gradle.test-retry"
             displayName = "Gradle test retry plugin"
             description = project.description
```
Comment (paplorinc) [inline plugin/build.gradle.kts:71]: my understanding is that `implementationClass` makes the `plugin/src/main/resources/META-INF/gradle-plugins/org.gradle.test-retry.properties` redundant, hence the duplicate warn - please validate

File: plugin/src/main/java/org/gradle/testretry/internal/executer/TestDescriptorImpl.java [hunk 4]
```diff
@@ -41,12 +41,6 @@ public Object getId() {
         return syntheticTestId;
     }
 
-    @Nullable
-    @Override
-    public Object getOwnerBuildOperationId() {
```
Comment (paplorinc) [inline plugin/src/main/java/org/gradle/testretry/internal/executer/TestDescriptorImpl.java:46]: removed in 7.0

File: plugin/src/test/groovy/org/gradle/testretry/testframework/SpockFuncTest.groovy [hunk 5]
```diff
@@ -853,7 +840,7 @@ class SpockFuncTest extends AbstractFrameworkFuncTest {
         return """
             dependencies {
                 implementation "org.codehaus.groovy:groovy-all:2.5.8"
-                testImplementation "org.spockframework:spock-core:1.3-groovy-2.5"
+                implementation "org.spockframework:spock-core:1.3-groovy-2.5"
```
Comment (paplorinc) [inline plugin/src/test/groovy/org/gradle/testretry/testframework/SpockFuncTest.groovy:843]: build gets confused otherwise

File: plugin/build.gradle.kts [hunk 6]
```diff
@@ -150,6 +153,7 @@ tasks.withType<Sign>().configureEach {
 
 tasks.withType<Test>().configureEach {
     maxParallelForks = 4
+    useJUnitPlatform()
```
Comment (paplorinc) [inline plugin/build.gradle.kts:156]: needed for upgraded Spock 2.0

File: plugin/src/test/groovy/org/gradle/testretry/testframework/SpockFuncTest.groovy [hunk 7]
```diff
@@ -670,22 +670,7 @@ class SpockFuncTest extends AbstractFrameworkFuncTest {
...
-
-            repositories {
-                mavenCentral()
-            }
-
-            dependencies {
-                implementation "org.codehaus.groovy:groovy-all:2.5.8", {
-                    exclude group: "org.junit.jupiter"
-                }
-                implementation "org.spockframework:spock-core:1.3-groovy-2.5"
```
Comment (paplorinc) [inline plugin/src/test/groovy/org/gradle/testretry/testframework/SpockFuncTest.groovy:682]: this was inherited by Spock 2 tests

File: plugin/src/test/groovy/org/gradle/testretry/testframework/SpockFuncTest.groovy [hunk 8]
```diff
@@ -304,25 +302,29 @@ class SpockFuncTest extends AbstractFrameworkFuncTest {
...
         then:
-        result.output.count('passingTest PASSED') == (isRerunsParameterizedMethods() ? 1 : 2)
+        result.output.with {
+            assert count('passingTest PASSED') == (isRerunsParameterizedMethods() ? 1 : 2)
 
-        result.output.count('unrolled[0] PASSED') == 2
-        result.output.count('unrolled[1] FAILED') == 2
-        result.output.count('unrolled[2] PASSED') == 2
+            readLines().with {
+                assert count { it =~ /unrolled ?\[.*?0] PASSED/ } == 2
```
Comment (paplorinc) [inline plugin/src/test/groovy/org/gradle/testretry/testframework/SpockFuncTest.groovy:309]: spock 1 displays `unrolled[0]` while Spock 2 adds the parameters to the name, like `unrolled [p1:v1, p2:v2 #0]`

File: buildSrc/src/main/groovy/org/gradle/testretry/build/PluginsVersionData.groovy [hunk 9]
```diff
@@ -1,10 +1,11 @@
 package org.gradle.testretry.build
 
+import groovy.xml.XmlSlurper
+
 class PluginsVersionData {
 
     static String latestVersion(String group, String name) {
         String url = "https://plugins.gradle.org/m2/${group.replace('.', '/')}/$name/maven-metadata.xml"
         return new XmlSlurper().parse(url).version
```
Comment (paplorinc) [inline buildSrc/src/main/groovy/org/gradle/testretry/build/PluginsVersionData.groovy:9]: default one is deprecated

File: gradle.properties [hunk 10]
```diff
@@ -1,6 +1,5 @@
 org.gradle.configureondemand=true
 org.gradle.parallel=true
-org.gradle.vfs.watch=true
```
Comment (marcphilipp) [inline gradle.properties:3]: 👍

File: plugin/build.gradle.kts (hunk 1 above)
Comment (marcphilipp) [reply plugin/build.gradle.kts:42]: I strongly suspect that happened by accident. I think we extract a local variable to remove the duplication and prevent it from happening again.

File: plugin/build.gradle.kts [hunk 11]
```diff
@@ -57,12 +58,14 @@ shadowJar.configure {
     }
 }
 
-tasks.getByName("jar").enabled = false
-tasks.getByName("jar").dependsOn(shadowJar)
+tasks.getByName("jar").apply {
```
Comment (marcphilipp) [inline plugin/build.gradle.kts:61]:
```suggestion
tasks.jar {
```

File: plugin/build.gradle.kts (hunk 2 above)
Comment (marcphilipp) [reply plugin/build.gradle.kts:68]: I don't think it makes a difference here. IIRC they are eagerly created regardless.

File: plugin/src/main/java/org/gradle/testretry/internal/config/TestRetryTaskExtensionAdapter.java [hunk 12]
```diff
@@ -38,28 +38,22 @@
...
-            : gradleVersion.getMajor() > 5;
-
-        this.useConventions = gradle51OrLater;
-
-        initialize(extension, gradle51OrLater);
+        initialize(extension, this.useConventions);
     }
 
-    private void initialize(TestRetryTaskExtension extension, boolean gradle51OrLater) {
+    private static void initialize(TestRetryTaskExtension extension, boolean gradle51OrLater) {
```
Comment (marcphilipp) [inline plugin/src/main/java/org/gradle/testretry/internal/config/TestRetryTaskExtensionAdapter.java:56]: I'd keep this non-static and reference the `useConventions` field directly instead of passing a boolean parameter.

File: plugin/src/main/java/org/gradle/testretry/internal/config/TestRetryTaskExtensionAdapter.java [hunk 13]
```diff
@@ -38,28 +38,22 @@
...
     public TestRetryTaskExtensionAdapter(
         ProviderFactory providerFactory,
         TestRetryTaskExtension extension,
         VersionNumber gradleVersion
-
     ) {
         this.providerFactory = providerFactory;
         this.extension = extension;
         this.simulateNotRetryableTest = Boolean.getBoolean(SIMULATE_NOT_RETRYABLE_PROPERTY);
+        this.useConventions = gradleVersion.compareTo(VersionNumber.version(5, 1)) >= 0;
```
Comment (marcphilipp) [inline plugin/src/main/java/org/gradle/testretry/internal/config/TestRetryTaskExtensionAdapter.java:51]: 👍

File: plugin/src/main/java/org/gradle/testretry/internal/config/TestTaskConfigurer.java [hunk 14]
```diff
@@ -65,10 +65,6 @@ private static void setTestExecuter(Test task, RetryTestExecuter retryTestExecut
         invoke(declaredMethod(Test.class, "setTestExecuter", TestExecuter.class), task, retryTestExecuter);
     }
 
-    private static boolean supportsPropertyConventions(VersionNumber gradleVersion) {
-        return gradleVersion.getMajor() == 5 ? gradleVersion.getMinor() >= 1 : gradleVersion.getMajor() > 5;
-    }
-
```
Comment (marcphilipp) [inline plugin/src/main/java/org/gradle/testretry/internal/config/TestTaskConfigurer.java:71]: Was this unused?

File: plugin/build.gradle.kts (hunk 3 above)
Comment (marcphilipp) [reply plugin/build.gradle.kts:71]: Indeed, the `java-gradle-plugin` does generate that file. Good catch! 👍

File: plugin/src/test/groovy/org/gradle/testretry/testframework/SpockFuncTest.groovy [hunk 15]
```diff
@@ -304,25 +302,29 @@ class SpockFuncTest extends AbstractFrameworkFuncTest {
         def result = gradleRunner(gradleVersion).buildAndFail()
 
         then:
-        result.output.count('passingTest PASSED') == (isRerunsParameterizedMethods() ? 1 : 2)
+        result.output.with {
+            assert count('passingTest PASSED') == (isRerunsParameterizedMethods() ? 1 : 2)
```
Comment (marcphilipp) [inline plugin/src/test/groovy/org/gradle/testretry/testframework/SpockFuncTest.groovy:306]:
```suggestion
        with(result.output) {
            count('passingTest PASSED') == (isRerunsParameterizedMethods() ? 1 : 2)
```

File: plugin/src/test/groovy/org/gradle/testretry/testframework/SpockFuncTest.groovy [hunk 16]
```diff
@@ -304,25 +302,29 @@ class SpockFuncTest extends AbstractFrameworkFuncTest {
...
 
-        result.output.count('unrolled[0] PASSED') == 2
-        result.output.count('unrolled[1] FAILED') == 2
-        result.output.count('unrolled[2] PASSED') == 2
+            readLines().with {
+                assert count { it =~ /unrolled ?\[.*?0] PASSED/ } == 2
+                assert count { it =~ /unrolled ?\[.*?1] FAILED/ } == 2
+                assert count { it =~ /unrolled ?\[.*?2] PASSED/ } == 2
+            }
```
Comment (marcphilipp) [inline plugin/src/test/groovy/org/gradle/testretry/testframework/SpockFuncTest.groovy:312]:
```suggestion
            with(readLines()) {
                count { it =~ /unrolled ?\[.*?0] PASSED/ } == 2
                count { it =~ /unrolled ?\[.*?1] FAILED/ } == 2
                count { it =~ /unrolled ?\[.*?2] PASSED/ } == 2
            }
```

File: plugin/src/test/groovy/org/gradle/testretry/testframework/SpockFuncTest.groovy [hunk 17]
```diff
@@ -304,25 +302,29 @@ class SpockFuncTest extends AbstractFrameworkFuncTest {
...
 
-        result.output.count('unrolled with param foo PASSED') == 2
-        result.output.count('unrolled with param bar FAILED') == 2
-        result.output.count('unrolled with param baz PASSED') == 2
+            assert count('unrolled with param foo PASSED') == 2
+            assert count('unrolled with param bar FAILED') == 2
+            assert count('unrolled with param baz PASSED') == 2
```
Comment (marcphilipp) [inline plugin/src/test/groovy/org/gradle/testretry/testframework/SpockFuncTest.groovy:316]:
```suggestion
            count('unrolled with param foo PASSED') == 2
            count('unrolled with param bar FAILED') == 2
            count('unrolled with param baz PASSED') == 2
```

File: plugin/src/test/groovy/org/gradle/testretry/testframework/SpockFuncTest.groovy [hunk 18]
```diff
@@ -304,25 +302,29 @@ class SpockFuncTest extends AbstractFrameworkFuncTest {
...
 
-        result.output.count('unrolled with param [foo] PASSED') == 2
-        result.output.count('unrolled with param [bar] FAILED') == 2
-        result.output.count('unrolled with param [baz] PASSED') == 2
+            assert count('unrolled with param [foo] PASSED') == 2
+            assert count('unrolled with param [bar] FAILED') == 2
+            assert count('unrolled with param [baz] PASSED') == 2
```
Comment (marcphilipp) [inline plugin/src/test/groovy/org/gradle/testretry/testframework/SpockFuncTest.groovy:320]:
```suggestion
            count('unrolled with param [foo] PASSED') == 2
            count('unrolled with param [bar] FAILED') == 2
            count('unrolled with param [baz] PASSED') == 2
```

File: plugin/src/test/groovy/org/gradle/testretry/testframework/SpockFuncTest.groovy (hunk 5 above)
Comment (marcphilipp) [reply plugin/src/test/groovy/org/gradle/testretry/testframework/SpockFuncTest.groovy:843]: ❓ How so?

File: plugin/src/test/groovy/org/gradle/testretry/testframework/SpockViaJUnitVintageFuncTest.groovy [hunk 19]
```diff
@@ -31,10 +31,8 @@ class SpockViaJUnitVintageFuncTest extends SpockBaseJunit5FuncTest {
     protected String buildConfiguration() {
         return """
             dependencies {
-                implementation "org.codehaus.groovy:groovy-all:2.5.8", {
-                    exclude group: "org.junit.jupiter"
-                }
-                testImplementation "org.spockframework:spock-core:1.3-groovy-2.5"
+                implementation "org.codehaus.groovy:groovy-all:2.5.8"
+                implementation "org.spockframework:spock-core:1.3-groovy-2.5"
```
Comment (marcphilipp) [inline plugin/src/test/groovy/org/gradle/testretry/testframework/SpockViaJUnitVintageFuncTest.groovy:35]: ❌ Why the change from `testImplementation` to `implementation`. That's not what users would do and hence not what we want to test here.

**Review by marcphilipp — CHANGES_REQUESTED** [review]
Thanks for doing this! The bulk of the changes looks great but we should get to the bottom of the `testImplementation` vs. `implementation` issue.

File: plugin/src/main/java/org/gradle/testretry/internal/config/TestRetryTaskExtensionAdapter.java (hunk 12 above)
Comment (paplorinc) [reply plugin/src/main/java/org/gradle/testretry/internal/config/TestRetryTaskExtensionAdapter.java:56]: That's called from the constructor, so it's important that it doesn't depend on the state yet

File: plugin/src/main/java/org/gradle/testretry/internal/config/TestTaskConfigurer.java (hunk 14 above)
Comment (paplorinc) [reply plugin/src/main/java/org/gradle/testretry/internal/config/TestTaskConfigurer.java:71]: yes

File: plugin/src/test/groovy/org/gradle/testretry/testframework/SpockFuncTest.groovy (hunk 15 above)
Comment (paplorinc) [reply plugin/src/test/groovy/org/gradle/testretry/testframework/SpockFuncTest.groovy:306]: ah, Spock 2 has that, nice!

File: plugin/src/test/groovy/org/gradle/testretry/testframework/SpockFuncTest.groovy (hunk 5 above)
Comment (paplorinc) [reply plugin/src/test/groovy/org/gradle/testretry/testframework/SpockFuncTest.groovy:843]: because the parent is in main, i.e. https://github.com/gradle/test-retry-gradle-plugin/blob/3578f5d29412e580e9b5d272ef2fc34b5f9161e0/plugin/src/test/groovy/org/gradle/testretry/testframework/SpockFuncTest.groovy#L673 - should I move that to a test folder as well?

File: plugin/src/test/groovy/org/gradle/testretry/testframework/SpockFuncTest.groovy (hunk 16 above)
Comment (paplorinc) [reply plugin/src/test/groovy/org/gradle/testretry/testframework/SpockFuncTest.groovy:312]: done

File: plugin/src/test/groovy/org/gradle/testretry/testframework/SpockFuncTest.groovy (hunk 17 above)
Comment (paplorinc) [reply plugin/src/test/groovy/org/gradle/testretry/testframework/SpockFuncTest.groovy:316]: done

File: plugin/src/test/groovy/org/gradle/testretry/testframework/SpockFuncTest.groovy (hunk 18 above)
Comment (paplorinc) [reply plugin/src/test/groovy/org/gradle/testretry/testframework/SpockFuncTest.groovy:320]: done

File: gradle/wrapper/gradle-wrapper.properties [hunk 20]
```diff
@@ -1,5 +1,5 @@
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-6.7-bin.zip
+distributionUrl=https\://services.gradle.org/distributions/gradle-7.0.2-bin.zip
```
[the remark this answers is not in the document]
Comment (paplorinc) [reply gradle/wrapper/gradle-wrapper.properties:3]: ran it multiple times, since I had to upgradle in smaller steps

File: samples/sample-junit5/build.gradle.kts [hunk 21]
```diff
@@ -9,9 +9,9 @@ repositories {
 }
 
 dependencies {
-    testImplementation("org.junit.jupiter:junit-jupiter-api:5.6.2")
-    testImplementation("org.junit.jupiter:junit-jupiter-params:5.6.2")
-    testRuntimeOnly("org.junit.jupiter:junit-jupiter-engine:5.6.2")
+    testImplementation("org.junit.jupiter:junit-jupiter-api:5.8.0-M1")
```
[the remark this answers is not in the document]
Comment (paplorinc) [reply samples/sample-junit5/build.gradle.kts:12]: no, but this is the latest version we're also using in dotcom - is needed for TD to fully function, so we should test with this, I think

File: plugin/build.gradle.kts (hunk 1 above)
Comment (paplorinc) [reply plugin/build.gradle.kts:42]: Done

File: plugin/src/test/groovy/org/gradle/testretry/testframework/SpockFuncTest.groovy (hunk 5 above)
Comment (marcphilipp) [reply plugin/src/test/groovy/org/gradle/testretry/testframework/SpockFuncTest.groovy:826]: So just for this one test? In that case, I think we should parameterize it or add an implementation dependency just for that test.

Comment (paplorinc) [reply plugin/src/test/groovy/org/gradle/testretry/testframework/SpockFuncTest.groovy:826]: I'll add a replace for this one test to change test dependencies to prod ones - at least this way other tests won't be affected

Comment (paplorinc) [reply plugin/src/test/groovy/org/gradle/testretry/testframework/SpockFuncTest.groovy]: Done

File: plugin/build.gradle.kts (hunk 11 above)
Comment (paplorinc) [reply plugin/build.gradle.kts]: Done

File: plugin/src/test/groovy/org/gradle/testretry/testframework/SpockViaJUnitVintageFuncTest.groovy (hunk 19 above)
Comment (paplorinc) [reply plugin/src/test/groovy/org/gradle/testretry/testframework/SpockViaJUnitVintageFuncTest.groovy]: Reverted

**Review by rpygithub — APPROVED** [review]

**Comment by paplorinc** [comment]
Still have to fix some cross version tests

**Review by marcphilipp — APPROVED** [review]

File: plugin/src/main/java/org/gradle/testretry/internal/config/TestRetryTaskExtensionAdapter.java (hunk 12 above)
Comment (marcphilipp) [reply plugin/src/main/java/org/gradle/testretry/internal/config/TestRetryTaskExtensionAdapter.java:60]: Ok, fair enough.

File: settings.gradle.kts [hunk 22]
```diff
@@ -6,7 +6,7 @@ pluginManagement {
 }
 
 plugins {
-    id("com.gradle.enterprise").version("3.5")
+    id("com.gradle.enterprise").version("3.6.1")
```
Comment (paplorinc) [inline settings.gradle.kts:9]:
```suggestion
    id("com.gradle.enterprise").version("3.6.2")
```

File: gradle/wrapper/gradle-wrapper.properties (hunk 20 above)
[the remark this answers is not in the document]
Comment (jonnybbb) [reply gradle/wrapper/gradle-wrapper.properties]: 👍 - ok was wondering why the wrapper jar did not change.

**Comment by paplorinc** [comment]
Thanks for the reviews!

File: gradle/wrapper/gradle-wrapper.properties (hunk 20 above)
[the remark this answers is not in the document]
Comment (paplorinc) [reply gradle/wrapper/gradle-wrapper.properties]: I did now, since I've bumped it to `7.1-rc-1`
